### DISCUSSION TITLE
Introduce `case-when` syntax.

### DISCUSSION THREAD

**Discussion by Kray-G** [Kray-G/kinx#177] [Archive]
I try to start discussion about some future features.

# Introduction

Thankfully according to increasing stars and visited by a lot of people, I am often asked about `break` in `switch-case`.
This is a design decision by following meanings, but I also sometimes think that I would like to solve this issue.

* Here are reasons.
  * This Kinx is to be similar with C and JavaScript.
  * The looking about no `break` in `switch-case` feels me the line will be connected to the next case. I imagine almost all C programmers will feel like that.

```c
switch (x) {
case 1:
    a = 1;
case 2:
    b = 1;
}
```

How do you think about it?

# Consideration

That is why, I considered the following 2 solutions.

* `break` should be necessary, but `fallthrough` keyword will be introduced and it is also necessary when it is fall through.
* Introducing another syntax like `case-when` instead of `switch-case`. `switch-case` will be stayed the same.

# Conclusion

I am now planning to introduce `case-when` syntax for this.
Although `switch-case` is a statement, I thnk to make `case-when` be as an expression.
Therefore, if you write multiple statements in `when` clause, I assume that you will have to write a callback function.
Maybe the arguments of that function will be values specified of  `case` and `when`.

```
var x = case y
    when 1:
        1
    when 2:
        10
    when variable: { &(y, variable)
        return y + 1;
    }
;
```

How about this?

Whatever you accept this or not, I am glad if giving me a feedback.
Of course I will be welcome if you have any other ideas.

**Comment by htsign** [comment]
I want the feature of `case-in` in ruby rather than `case-when`.
If adopted this, we can write a code like below:

```javascript
var x = case [y, z]
    when [1, _]: 1
    when [2, _]: 10
    when [v, 0]: {
        return v + 42;
    }
    default: // or `else` keyword
        0
;
```

**Reply by Kray-G** [reply]
Thanks for the suggestion!

You mean it is a pattern matching syntax, right?
I feel the meaning between `case-when` and `case-in` is different, so I think it means it is necessary to support both syntaxes.
If so, in that case, should it be written as `case-in` instead of `case-when`?
Or is it possible to use the same syntax for the above 2 meanings... I will try to consider it.

If my understanding is correct, the last `when` case means, if `z` equals 0, then the value of `y` is assigned to the variable of `v`.
As a result, `return v + 42` means it is same as `return y + 42`, right?

**Reply by htsign** [reply]
> If my understanding is correct, the last when case means, if z equals 0, then the value of y is assigned to the variable of v.
As a result, return v + 42 means it is same as return y + 42, right?

Yes, it is what I want to say.
And I think `case-in` in ruby is compatible with `case-when`.

```javascript
var x = case [y, z]
    when [1, _]: 1
    when [v, _]: // match any cases
        v + 42
    ...
```
```javascript
var x = case [y]
    when [1]: 1
    when [v]: // match any cases
        v + 42
    ...
```
```javascript
var x = case y
    when 1: 1
    when v: // match any cases
        v + 42
    ...
```

**Reply by Kray-G** [reply]
Thanks for examples, I understood what you meant. But I have a question about one thing. If I have misunderstanding, please let me know.

At the last case, I think the variable `v` is not an lvalue, don't you think so?
It is because it means "when y equals to v". Therefore it doesn't mean it matches to any cases, and I think it means, if `v` has some value and `y` equals to that value, it will return `v + 42`.

What do you think?

Thanks in advance.

**Reply by htsign** [reply]
> Therefore it doesn't mean it matches to any cases, and I think it means, if v has some value and y equals to that value, it will return v + 42.

Oh, that's true.
During the discussion, I equated it with the pattern matching - it used in the functional paradigm.
By the way, other languages solve the problem by adding expressions:

```ocaml
(* OCaml *)
let x =
  match y with
  | 1 -> 1
  | 2 -> 10
  (* `=` is equality operator / it means "y is captured in v and v equals z then ..." *)
  | v when v = z -> v + 42
  (* same above *)
  | _ when y = z -> y + 42
  (* default case *)
  | _ -> -1
```

```scala
// Scala
val x = y match {
  case 1 => 1
  case 2 => 10
  case v if v == z => v + 42
  case _ => -1
}
```

**Reply by Kray-G** [reply]
I see, you mean it is possible to use `if` modifier for the special condition, right?
Surely, it looks the case of a pattern matching syntax will include a normal `case-when` in Ruby...

Here are some examples. I think the range will be also available.
How about these?

```
var x = case y
    when 1:
        1
    when 2..10:
        10
    when v if v == m:
        v + 42
    when v if v == n:
        v + 58
    else:
        -1
;
```

For array, we can use a spread/rest operator.

```
var x = case [y, z, i, j, k]
    when [1, ...a]:
        a[2]
    when [2..10, a, b]:
        a + b + 10
    when [v, m, ...n] if 10 < v && v <= 20:
        v + m + n[0] + 42
    when [v, 50, ...a] if v == n:
        v + 58
    else:
        -1
;
```

For object, Kinx don't have a spread/rest operator.

```
var obj = { x: 10, y: 20, sub: { a: 1, b: 2, c: 3 } };
var x = case obj
    when { x: 5, y: y, sub: z } if y == m:
        { => y }
    when { x: 6, y: y, sub: { a: 1, b: b, c: c} }:
        y + b + c;
    when { x: 10, y: y, sub: z }:
        10
;
```

Thanks and best regards.

**Reply by htsign** [reply]
Great examples. I love that codes!! 😄
But it's hard to implement them, don't you?

**Reply by Kray-G** [reply]
Okay I will make it!
I'm going to define a specification clearly first based on this discussion.
As you say the implementation may be hard, so I will try some ideas.

Many thanks a lot.

**Comment by Kray-G** [comment]
# Conclusion

## Overview

I have implemented this feature as suited with a current supported feature,
but there are some differences from the examples on this discussion.

1. If-modifier ... condition needs parentheses like `if (v == m)`
2. Colon ... you don't need to put a colon. It is possible to put a colon to separate a condition and a body of `when` clause, but it is just for an appearance.
3. Block & function ... by putting a block to `when` clause it is an automatically called function object. But this is a special feature for a block. A function like `{ => y }` which is not a block is never called automatically. A function object itself is just returned as a result of `when` clause. When you want to call it, use `{ => y}()` instead.
4. Semicolon ... `case-when` is an expression, so a body of `when` clause except in a block can not use a semicolon.

## Related Updates

Before I implemented this feature, I tried that a pattern matching mechanism is applied to declaration statement first.
It is because I thought that this feature can be realized by the same mechanism.
That is correct and it was successful to implement it.
Therefore the pattern matching algorithm now can be used also in the following syntaxes.

* declaration.
* assignment.
* function argument.

See documents described in References section for details.

## References

Finally, I wrote a specification for this feature at a document below.

* [case-when](https://kray-g.github.io/kinx/spec/statement/expression/case_when.html)
    * A specification of `case-when`.
* [declaration](https://kray-g.github.io/kinx/spec/statement/declaration.html)
    * It supported a pattern matching mechanism also in declaration.
* [assignment](https://kray-g.github.io/kinx/spec/statement/expression/assign.html)
    * It supported a pattern matching mechanism also in assignment.
* [function](https://kray-g.github.io/kinx/spec/definition/function.html)
    * It supported a pattern matching mechanism also in function arguments.

If you have any problem, please let me know.

Thank you very much.

**Comment by Kray-G** [comment]
I updated `case-when` and some related features in the current master trunk.
It will be released in the next release.
See [switch-when](spec/statement/switch_when.md) and [Case-When Expression](expression/case_when.md) for details.
The summary of those are as follows.

* Updates of `case-when`
    * Pin operator such as `^a`. Just `a` means an lvalue to match all, but pin operator of `^a` means the value is a rvalue and to compare values.
    * Alternative pattern by `|`. `1 | 2 | 3` means to match `1`, to match `2`, or to match `3`.
    * Error when it is array and the length is different.
    * Putting a function object in a `when` clause is available. It means to check if a returned value by the function is true.
    * `_` in the `when` clause means to ignore binding a value at `when` condition.
* Introduced `switch-when` statement.
    * The benefit of `switch-case` is to compile to a jump by table in some cases. This is good for the performance, but it never do so in `case-when`.
    * It is possible to use `when` instead of `case` in `switch` statement. When `when` is used, `break` will be automatically done after the last statement.
    * Instead, `fallthrough` keyword is prepared and it will make it disable to do auto break.
    * `else` label is also prepared. It means a `default` label with auto break.
    * Of course,  `switch-when` statement can be also used in `native`.

If you have any problem, feel free to let me know.
Best regards.

**Reply by htsign** [reply]
Great update!

> Putting a function object in a `when` clause is available. It means to check if a returned value by the function is true.

What is meaning of "true", `True` or `true` that is an alias of `1` ?
As far as I read "Function Call in When Clause" section of the reference, I think it's the latter.

> `It is possible to use `when...

It is wrong the markup.

**Reply by Kray-G** [reply]
Thanks for the feedback!

> > Putting a function object in a `when` clause is available. It means to check if a returned value by the function is true.
>
> What is meaning of "true", `True` or `true` that is an alias of `1` ?
> As far as I read "Function Call in When Clause" section of the reference, I think it's the latter.

Now in the current code, the `true` means just a **non-zero** value, but it will be changed a little in the release,
because I feel it had better be same as a condition of `if`. This means the `a` of `if (a)`.
To say it a little more detailed, it will meet the following conditions.

* `undefined`/`null` ... always **false**.
* Integer ... **true** means a **non-zero** value.
* BigInteger ... always **true**.
* Double ... **true** if the distance between 2 values are **bigger than DBL_EPSILON**.
* String ... **true** if it is **not empty string**.
* Function ... always **false**.
* Object ... basically **true**.

> > `It is possible to use `when...
>
> It is wrong the markup.

Thanks for a report. I fixed it.

best regards.
